buildingmotif/shape_builder/shape.py

```diff
@@ -0,0 +1,168 @@
+from typing import List, Optional, Tuple, Union
+
+from rdflib import RDF, BNode, Graph, Literal, URIRef
+from rdflib.collection import Collection
+from rdflib.term import Node
+
+from buildingmotif.namespaces import SH, A, bind_prefixes
+
+
+class Shape(Graph):
+    def __init__(
+        self,
+        identifier: Optional[Union[Node, str]] = None,
+        message: Optional[str] = None,
+    ) -> None:
+        super().__init__(identifier=identifier)
+        bind_prefixes(self)
+
+        if message:
+            self.add((self.identifier, SH["message"], Literal(message)))
+
+    def add_property(self, property: URIRef, object: Node):
+        self.add((self, property, object))
+
+        return self
+
+    def add_list_property(
+        self, property: URIRef, nodes: Union[List[Node], Tuple[Node, ...]]
+    ):
+        identifier = BNode()
+        Collection(self, identifier, nodes)
+
+        self.add((self, property, identifier))
+
+        return self
+
+    def OR(self, *nodes: Node):
+        self.add_list_property(SH["or"], nodes)
+        return self
+
+    def AND(self, *nodes: Node):
+        self.add_list_property(SH["and"], nodes)
+        return self
+
+    def NOT(self, node: Node):
+        self.add((self, SH["not"], node))
+
+        return self
+
+    def XONE(self, *nodes: Node):
+        self.add_list_property(SH["xone"], nodes)
+        return self
+
+    def add(self, triple: Tuple[Node, Node, Node]):
+        (s, p, o) = triple
+        if isinstance(o, Shape):
+            self += o
+            o = o.identifier
+        if s == self:
+            s = s.identifier
+        triple = (s, p, o)
+        return super().add(triple)
+
+
+class NodeShape(Shape):
+    def __init__(
+        self,
+        identifier: Optional[Union[Node, str]] = None,
+        message: Optional[str] = None,
+    ) -> None:
+        super().__init__(identifier=identifier, message=message)
+
+        self.add((self, A, SH["NodeShape"]))
+
+    def of_class(self, class_: Node, target=False):
+        predicate = SH["class"]
+        if target:
+            predicate = SH["target_class"]
+
+        self.add((self, predicate, class_))
+
+        return self
+
+    def has_property(self, property: Node = None):
+        if isinstance(property, URIRef):
+            property = PropertyShape().has_path(property)
+
+        self.add((self, SH["property"], property))
+
+        return self
+
+
+class PropertyShape(Shape):
```

Comment:
I think I lean toward explicit qualification declaration for now. Since this is a lower level shape building API I don't think it's unreasonable to expect that people who use it have a basic understanding of qualified vs universal (though then again I obviously missed something).